### PR TITLE
Standardize usage of the 'Maslow' string

### PR DESCRIPTION
**Pull request by aersida**
Standardizes (standardises) most usages of "Maslow" in various strings.

~Also standardises usage of `&somestring[0]` instead of `somestring.c_str()` when converting from `std::string` to `char *`~ - rolled this back out

### CODE REVIEW

**Comment by davidelang** [comment]
@md8n This crossed with a PR I was creating (which I have now closed)

could you make an added commit to this PR that takes the two private variables _beltEndExtension and _armLength and exposes them through the config file (while giving them their current defaults if they are not defined in the file)?

I think this is moving them up a couple lines from the private section in Maslow.h and then tying them to config variable names in MachineConfig.cpp and ProcessSettings.cpp I don't know if they can be set to a default the way they are currently in Maslow.h or if added logic needs to be put in one of the other files.

I also want to add a variable for "stock thickness" that I will then add in the 4x computXX() functions and the measurementToXYPlane() function in Maslow.cpp (adding the variable to the Z calculation). It would need added thinking to figure where to add it to the UI, I want to add support in the firmware first.

**Comment by davidelang** [comment]
later on, I would like to see if we can change the internal processing from separate variables (trZ, tlZ, brZ, blZ) to arrays (for example, zOffset[arm] where x is 0-3 with tl=0, tr=1, bl=2, br=3)

We already have quite a few places where we have the exact same thing being done to each arm, where the only change is the variable/function names used. Making it so that we can collapse that code into a single instance with the arm as a parameter will simplify the codebase by eliminating functions or turning things into a small loop (and if we want to keep things as 4 lines for the 4 arms, using the names instead of the numbers for the measurements array will help make it clearer what is happening)

Going forward it will make it easier to add per-arm changes such as stops/extensions. As I see it we either need to make the variables into arrays, or we will have to change measurementToXYPlane into four functions, with the only difference being the arm they act on.

But this is a fairly major overhaul of Maslow.cpp, so we should do the other work first

**Comment by aersida** [comment]
> @md8n This crossed with a PR I was creating (which I have now closed)
> 

My plan for this is purely to be a cleanup PR before doing such things. That's really my motivation here.

**Comment by aersida** [comment]
> later on, I would like to see if we can change the internal processing from separate variables (trZ, tlZ, brZ, blZ) to arrays (for example, zOffset[arm] where x is 0-3 with tl=0, tr=1, bl=2, br=3)
> 
> We already have quite a few places where we have the exact same thing being done to each arm, where the only change is the variable/function names used. Making it so that we can collapse that code into a single instance with the arm as a parameter will simplify the codebase by eliminating functions or turning things into a small loop 
> 
> But this is a fairly major overhaul of Maslow.cpp, so we should do the other work first

Yep - This PR is really about the housekeeping necessary before doing such things

**Comment by davidelang** [comment]
I agree, I'm just asking if you could do a separate commit/PR for the two variables that builds on this one rather than me trying to build on it (both because I think you understand this code better, so are less likely to make mistakes, and also because it's less confusing to have two PRs from one person where one builds on the other than having them come from separate people)

**Comment by darrenhearn** [comment]
I'm all for standardi[zs]ing :-), but am curious about the preference for `&str[0]` over `str.c_str()`. My understanding is that `str.c_str()` guarantees null termination (but `&str[0]` does not). While I don't think there are null termination risks for the specific changes in this PR, my interest is more about potential future changes where there *is* a null termination risk (presumably accidental). In that case, adopting the `str.c_str()` convention could save us from some sneaky bugs/painful troubleshooting. Thoughts?

**Comment by BarbourSmith** [comment]
This looks great to me. I'm curious about @darrenhearn's ideas, so I'll hold off on merging it until we think it thorugh, but either option seems like an improvement and excellent work! 👍 👍 

Great cleanup! This kind of thing can go neglected easily, so extra kudos for being on top of it 😍

**Comment by aersida** [comment]
> I'm all for standardi[zs]ing :-), but am curious about the preference for `&str[0]` over `str.c_str()`. My understanding is that `str.c_str()` guarantees null termination (but `&str[0]` does not). While I don't think there are null termination risks for the specific changes in this PR, my interest is more about potential future changes where there _is_ a null termination risk (presumably accidental). In that case, adopting the `str.c_str()` convention could save us from some sneaky bugs/painful troubleshooting. Thoughts?

The real issue is avoiding the usage of `c_str()`, and there are other options for doping that. In this case I chose to use `&str[0]` because it's 'the oldest' technique for doing this, but with that caveat of the null termination problem. I'd have no issue with trying one of the other options, I would doubt there's any risk for us of having to work with really old C++ compilers.

And why not use `c_str()`? It copies the string, so it leads to additional memory consumption. In some cases (if as programmers we've really messed up) it can lead to memory leaks. So, it's considered as 'not the best choice' for converting from `std::string` to `char *`

**Comment by aersida** [comment]
> This looks great to me. I'm curious about @darrenhearn's ideas, so I'll hold off on merging it until we think it thorugh, but either option seems like an improvement and excellent work! 👍 👍
> 
> Great cleanup! This kind of thing can go neglected easily, so extra kudos for being on top of it 😍

I could see where development was heading (hi @davidelang ) and I knew from past poking around this bit of code that we had more than our fair share of 'magic strings' in there. Therefore, my thinking went to "it would probably be a good idea to do some housekeeping before we start moving the furniture around too much"

But I'd definitely agree about holding off on merging. There's a corresponding change required on the front end code. With this PR I cleaned up some usage of `'maslow'` vs. `'Maslow'` and the existing front end code will break on that.

**Comment by davidelang** [comment]
would it make sense to split the maslow/Maslow cleanup from the c_str() cleanup?

**Comment by darrenhearn** [comment]
> > I'm all for standardi[zs]ing :-), but am curious about the preference for `&str[0]` over `str.c_str()`. My understanding is that `str.c_str()` guarantees null termination (but `&str[0]` does not). While I don't think there are null termination risks for the specific changes in this PR, my interest is more about potential future changes where there _is_ a null termination risk (presumably accidental). In that case, adopting the `str.c_str()` convention could save us from some sneaky bugs/painful troubleshooting. Thoughts?
> 
> The real issue is avoiding the usage of `c_str()`, and there are other options for doping that. In this case I chose to use `&str[0]` because it's 'the oldest' technique for doing this, but with that caveat of the null termination problem. I'd have no issue with trying one of the other options, I would doubt there's any risk for us of having to work with really old C++ compilers.
> 
> And why not use `c_str()`? It copies the string, so it leads to additional memory consumption. In some cases (if as programmers we've really messed up) it can lead to memory leaks. So, it's considered as 'not the best choice' for converting from `std::string` to `char *`

I recognize that I don't have any standing to nitpick (though hopefully I'll finish work for a PR in the next week to establish some "street cred" :-), but my understanding since C++11 is that `c_str()` does *not* perform a copy/consume additional memory. Instead, it returns a `const char*` to the array that backs the `std::string` object (with a null-termination guarantee). This is borne out in the `c_str()` [docs for GNU libstdc++](https://gcc.gnu.org/onlinedocs/gcc-12.4.0/libstdc++/api/a05380.html#ac0b617aa703b1cb8bcd1f806a7caa0fe) and implementation (for both `std::string`, in `basic_string.h` from PlatformIO's copy of GNU libstdc++ for the ESP32, and the Arduino `String`, in `WString.h` from PlatformIO's copy of the ESP32 Arduino framework). As you noted, we probably don't have to worry about compatibility with older compilers/standards (pre-C++11); in particular, the current `platformio.ini` includes a build flag for `-std=gnu++17`. So, unless I'm missing something, I don't think the copy/memory concern is a reason to avoid `c_str()`.

One other point is that FluidNC already leaned heavily towards the `c_str()` convention; so adopting `&str[0]` (or alternatives) within the Maslow subset of code means we have two different conventions.
```
$ grep -R 'c_str()' FluidNC/* | wc -l
203

$ grep -R -e '&[_a-zA-Z].*\[0\]' FluidNC/* | wc -l
2
```

Having said all that, I swear I'm not a `c_str()` zealot! I'm happy to work with whatever convention you Maslow veterans recommend; I just want to make sure it's a fully "eyes wide open" decision.

**Comment by aersida** [comment]
> > > I'm all for standardi[zs]ing :-), but am curious about the preference for `&str[0]` over `str.c_str()`. My understanding is that `str.c_str()` guarantees null termination (but `&str[0]` does not). While I don't think there are null termination risks for the specific changes in this PR, my interest is more about potential future changes where there _is_ a null termination risk (presumably accidental). In that case, adopting the `str.c_str()` convention could save us from some sneaky bugs/painful troubleshooting. Thoughts?
> > 
> > 
> > The real issue is avoiding the usage of `c_str()`, and there are other options for doping that. In this case I chose to use `&str[0]` because it's 'the oldest' technique for doing this, but with that caveat of the null termination problem. I'd have no issue with trying one of the other options, I would doubt there's any risk for us of having to work with really old C++ compilers.
> > And why not use `c_str()`? It copies the string, so it leads to additional memory consumption. In some cases (if as programmers we've really messed up) it can lead to memory leaks. So, it's considered as 'not the best choice' for converting from `std::string` to `char *`
> 
> I recognize that I don't have any standing to nitpick (though hopefully I'll finish work for a PR in the next week to establish some "street cred" :-), but my understanding since C++11 is that `c_str()` does _not_ perform a copy/consume additional memory. Instead, it returns a `const char*` to the array that backs the `std::string` object (with a null-termination guarantee). This is borne out in the `c_str()` [docs for GNU libstdc++](https://gcc.gnu.org/onlinedocs/gcc-12.4.0/libstdc++/api/a05380.html#ac0b617aa703b1cb8bcd1f806a7caa0fe) and implementation (for both `std::string`, in `basic_string.h` from PlatformIO's copy of GNU libstdc++ for the ESP32, and the Arduino `String`, in `WString.h` from PlatformIO's copy of the ESP32 Arduino framework). As you noted, we probably don't have to worry about compatibility with older compilers/standards (pre-C++11); in particular, the current `platformio.ini` includes a build flag for `-std=gnu++17`. So, unless I'm missing something, I don't think the copy/memory concern is a reason to avoid `c_str()`.
> 
> One other point is that FluidNC already leaned heavily towards the `c_str()` convention; so adopting `&str[0]` (or alternatives) within the Maslow subset of code means we have two different conventions.
> 
> ```
> $ grep -R 'c_str()' FluidNC/* | wc -l
> 203
> 
> $ grep -R -e '&[_a-zA-Z].*\[0\]' FluidNC/* | wc -l
> 2
> ```
> 
> Having said all that, I swear I'm not a `c_str()` zealot! I'm happy to work with whatever convention you Maslow veterans recommend; I just want to make sure it's a fully "eyes wide open" decision.

Which serves me right for reading stackoverflow messages that are too old - I'll change it all to `c_str()` which should reduce the noise of this PR

**Review by BarbourSmith — APPROVED** [review]
This looks like an excellent clean up to me 👍 👍 

Any objection to merging? Are we still waiting on a UI change that complements it?

File: FluidNC/src/Machine/MachineConfig.cpp [hunk 1]
```diff
@@ -94,36 +94,36 @@ namespace Machine {
...
-        handler.item("Maslow_brX", Maslow.brX);
-        handler.item("Maslow_brY", Maslow.brY);
+        handler.item(M+"_brX", Maslow.brX);
+        handler.item(M+"_brY", Maslow.brY);
 
-        handler.item("Maslow_tlZ", Maslow.tlZ);
-        handler.item("Maslow_trZ", Maslow.trZ);
-        handler.item("Maslow_blZ", Maslow.blZ);
-        handler.item("Maslow_brZ", Maslow.brZ);
+        handler.item(M+"_tlZ", Maslow.tlZ);
```
Comment (BarbourSmith) [inline FluidNC/src/Machine/MachineConfig.cpp:114]: I kind of imagined that the compiler would optimize out these strings, is that not the case?

**Review by BarbourSmith — APPROVED** [review]
Fantastic! This is great work!

**Comment by davidelang** [comment]
BarbourSmith wrote:

> -        handler.item("Maslow_tlZ", Maslow.tlZ);
> -        handler.item("Maslow_trZ", Maslow.trZ);
> -        handler.item("Maslow_blZ", Maslow.blZ);
> -        handler.item("Maslow_brZ", Maslow.brZ);
> +        handler.item(M+"_tlZ", Maslow.tlZ);
>
> I kind of imagined that the compiler would optimize out these strings, is that not the case?

As I see it, this is doing two things

1. making sure that everything is consistant
2. make it so that if there is ever a reason to change the string, there is only 
one place to make the change.

David Lang

**Pull request merged**